File: ros_ws/src/ros_python/ros_python/service_client.py
from geometry_msgs.msg import Twist
from custom_interface.srv import MoveMessage
import rclpy
from rclpy.node import Node
import sys

class ServiceClient(Node):
    def __init__(self):
        super().__init__("service_client")
        self.client = self.create_client(MoveMessage, 'move')
        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('service not available, waiting...')
        self.req = MoveMessage.Request()
    def send_request(self):
        self.req.move = "right" #sys.argv[1]
        self.future = self.client.call_async(self.req)

def main(args=None):
    rclpy.init(args=args)
    client = ServiceClient()
    client.send_request()
    while rclpy.ok():
        rclpy.spin_once(client)
        if client.future.done():
            try:
                response = client.future.result()
            except Exception as e:
                client.get_logger().info('service call failed %r'%(e,))
            else:
                client.get_logger().info('Response state: %r'%(response.success,))
            break
    client.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: ros_ws/src/cpp/src/service_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_srvs/srv/empty.hpp"
#include <chrono>
//#include <cstdLib> 
#include <memory>

using namespace std::chrono_literals;

int main(int argc, char* argv[]){
    rclcpp::init(argc, argv);
    std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("service_client");
    rclcpp::Client<std_srvs::srv::Empty>::SharedPtr client = node->create_client<std_srvs::srv::Empty>("moving_right");
 
    auto request = std::make_shared<std_srvs::srv::Empty::Request>();

    while (!client->wait_for_service(std::chrono::seconds(1))){
        if (!rclcpp::ok()){
            RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Interrupted while waiting for the service. Exiting.");
            return 0;
        }
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Service not available, waiting again...");
    }
    auto result_future = client->async_send_request(request);
    if (rclcpp::spin_until_future_complete(node, result_future) == rclcpp::FutureReturnCode::SUCCESS){
        auto result = result_future.get();
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "The Robot is moving.");
    }else{
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Failed to call service /moving");
    }
    rclcpp::shutdown();
    return 0;
}